external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/HdfsSpout.java

```diff
@@ -0,0 +1,739 @@
+
+  private Configuration hdfsConfig;
+
+  private Map conf = null;
+  private FileLock lock;
+  private String spoutId = null;
+
+  HdfsUtils.Pair<Path,FileLock.LogEntry> lastExpiredLock = null;
+  private long lastExpiredLockTime = 0;
+
+  private long tupleCounter = 0;
+  private boolean ackEnabled = false;
+  private int acksSinceLastCommit = 0 ;
+  private final AtomicBoolean commitTimeElapsed = new AtomicBoolean(false);
+  private Timer commitTimer;
+  private boolean fileReadCompletely = true;
+
+  private String configKey = Configs.DEFAULT_HDFS_CONFIG_KEY; // key for hdfs Kerberos configs
+
+  public HdfsSpout() {
+  }
+  /** Name of the output field names. Number of fields depends upon the reader type */
+  public HdfsSpout withOutputFields(String... fields) {
+    outputFields = new Fields(fields);
+    return this;
+  }
+
+  /** set key name under which HDFS options are placed. (similar to HDFS bolt).
+   * default key name is 'hdfs.config' */
+  public HdfsSpout withConfigKey(String configKey) {
+    this.configKey = configKey;
+    return this;
+  }
+
+  public Path getLockDirPath() {
+    return lockDirPath;
+  }
+
+  public SpoutOutputCollector getCollector() {
+    return collector;
+  }
+
+  public void nextTuple() {
+    LOG.trace("Next Tuple {}", spoutId);
+    // 1) First re-emit any previously failed tuples (from retryList)
+    if (!retryList.isEmpty()) {
+      LOG.debug("Sending tuple from retry list");
+      HdfsUtils.Pair<MessageId, List<Object>> pair = retryList.remove();
+      emitData(pair.getValue(), pair.getKey());
+      return;
+    }
+
+    if( ackEnabled  &&  tracker.size()>= maxOutstanding) {
+      LOG.warn("Waiting for more ACKs before generating new tuples. " +
+              "Progress tracker size has reached limit {}, SpoutID {}"
+              , maxOutstanding, spoutId);
+      // Don't emit anything .. allow configured spout wait strategy to kick in
+      return;
+    }
+
+    // 2) If no failed tuples to be retried, then send tuples from hdfs
+    while (true) {
+      try {
+        // 3) Select a new file if one is not open already
+        if (reader == null) {
+          reader = pickNextFile();
+          if (reader == null) {
+            LOG.debug("Currently no new files to process under : " + sourceDirPath);
+            return;
+          } else {
+            fileReadCompletely=false;
+          }
+        }
+        if( fileReadCompletely ) { // wait for more ACKs before proceeding
+          return;
+        }
+        // 4) Read record from file, emit to collector and record progress
+        List<Object> tuple = reader.next();
+        if (tuple != null) {
+          fileReadCompletely= false;
+          ++tupleCounter;
+          MessageId msgId = new MessageId(tupleCounter, reader.getFilePath(), reader.getFileOffset());
+          emitData(tuple, msgId);
+
+          if(!ackEnabled) {
```

Comment:
If acks are disabled by storm ack is called as part of the call to emit.  I think you may ahve extra logic in here that you do not need.